{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from dammit.fileio.gff3 import GFF3Parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'assemblies/'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-7ecb360f438f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0massemblies_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"assemblies/\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0massemblies\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlistdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0massemblies_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mannotations_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"gff3/\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mgff3\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlistdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannotations_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgff3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'assemblies/'"
     ]
    }
   ],
   "source": [
    "assemblies_dir = \"assemblies/\"\n",
    "assemblies = os.listdir(assemblies_dir)\n",
    "annotations_dir = \"gff3/\"\n",
    "gff3 = os.listdir(annotations_dir)\n",
    "print(gff3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['F_sciadicus.trinity_out.Trinity.fasta', 'F_grandis.trinity_out.Trinity.fasta', 'F_catanatus.trinity_out.Trinity.fasta', 'L_goodei.trinity_out.Trinity.fasta', 'F_parvapinis.trinity_out.Trinity.fasta', 'F_zebrinus.trinity_out.Trinity.fasta', 'F_heteroclitusMDPL.trinity_out.Trinity.fasta', 'F_heteroclitusMDPP.trinity_out.Trinity.fasta', 'F_notatus.trinity_out.Trinity.fasta', 'F_rathbuni.trinity_out.Trinity.fasta', 'F_diaphanus.trinity_out.Trinity.fasta', 'F_notti.trinity_out.Trinity.fasta', 'F_chrysotus.trinity_out.Trinity.fasta', 'A_xenica.trinity_out.Trinity.fasta', 'L_parva.trinity_out.Trinity.fasta', 'F_similis.trinity_out.Trinity.fasta', 'F_olivaceous.trinity_out.Trinity.fasta']\n"
     ]
    }
   ],
   "source": [
    "print(assemblies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/dammit_working_gff3/lib/python3.6/site-packages/dammit-1.1-py3.6.egg/dammit/fileio/base.py:79: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  return pd.concat(self, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "#annotations = GFF3Parser(filename=\"gff3/F_notatus.trinity_out.Trinity.fasta.dammit.gff3\").read()\n",
    "annotations = GFF3Parser(filename=\"gff3/A_xen_test.gff3\").read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>0 100 +</th>\n",
       "      <th>0 1000 +</th>\n",
       "      <th>0 1002 +</th>\n",
       "      <th>0 1004 +</th>\n",
       "      <th>0 1005 +</th>\n",
       "      <th>0 1006 +</th>\n",
       "      <th>0 1007 +</th>\n",
       "      <th>0 1008 +</th>\n",
       "      <th>0 101 +</th>\n",
       "      <th>...</th>\n",
       "      <th>score</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>trunc</th>\n",
       "      <th>type_x</th>\n",
       "      <th>type_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500000e-26</td>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6.200000e-15</td>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2.600000e-08</td>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.200000e-62</td>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>LAST</td>\n",
       "      <td>247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.800000e-142</td>\n",
       "      <td>Transcript_0</td>\n",
       "      <td>LAST</td>\n",
       "      <td>247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 15599 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       0 100 + 0 1000 + 0 1002 + 0 1004 + 0 1005 + 0 1006 + 0 1007 + 0 1008 +  \\\n",
       "0  NaN     NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "1  NaN     NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "2  NaN     NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "3  NaN     NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "4  NaN     NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "\n",
       "  0 101 +  ...            score         seqid source start strand strand_x  \\\n",
       "0     NaN  ...     5.500000e-26  Transcript_0  HMMER    18    NaN      NaN   \n",
       "1     NaN  ...     6.200000e-15  Transcript_0  HMMER    18    NaN      NaN   \n",
       "2     NaN  ...     2.600000e-08  Transcript_0  HMMER    66    NaN      NaN   \n",
       "3     NaN  ...     3.200000e-62  Transcript_0   LAST   247    NaN        +   \n",
       "4     NaN  ...    3.800000e-142  Transcript_0   LAST   247    NaN        +   \n",
       "\n",
       "  strand_y trunc                       type_x type_y  \n",
       "0      NaN   NaN            protein_hmm_match    NaN  \n",
       "1      NaN   NaN            protein_hmm_match    NaN  \n",
       "2      NaN   NaN            protein_hmm_match    NaN  \n",
       "3      NaN   NaN  translated_nucleotide_match    NaN  \n",
       "4      NaN   NaN  translated_nucleotide_match    NaN  \n",
       "\n",
       "[5 rows x 15599 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# dammit\r\n",
      "## a tool for easy de novo transcriptome annotation\r\n",
      "\r\n",
      "by Camille Scott\r\n",
      "\r\n",
      "**v1.1**, 2018\r\n",
      "\r\n",
      "usage: dammit [-h] [--debug] [--version] {migrate,databases,annotate} ...\r\n",
      "\r\n",
      "# dammit: a tool for easy de novo transcriptome annotation\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "  --debug\r\n",
      "  --version             show program's version number and exit\r\n",
      "\r\n",
      "dammit subcommands:\r\n",
      "  {migrate,databases,annotate}\r\n",
      "    databases           Check for databases and optionally download and\r\n",
      "                        prepare them for use. By default, only check their\r\n",
      "                        status.\r\n",
      "    annotate            The main annotation pipeline. Calculates assembly\r\n",
      "                        stats; runs BUSCO; runs LAST against OrthoDB (and\r\n",
      "                        optionally uniref90), HMMER against Pfam, Inferal\r\n",
      "                        against Rfam, and Conditional Reciprocal Best-hit\r\n",
      "                        Blast against user databases; and aggregates all\r\n",
      "                        results in a properly formatted GFF3 file.\r\n"
     ]
    }
   ],
   "source": [
    "!dammit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.23.4'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A_xenica.trinity_out.Trinity.fasta.dammit.gff3\tgff3\r\n",
      "assemblies\t\t\t\t\tnamemaps\r\n",
      "check_annotations_fasta.ipynb\r\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations = GFF3Parser(filename=\"A_xenica.trinity_out.Trinity.fasta.dammit.gff3\").read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Target</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>database</th>\n",
       "      <th>end</th>\n",
       "      <th>env_coords</th>\n",
       "      <th>phase</th>\n",
       "      <th>score</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>trunc</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>homology:425062</td>\n",
       "      <td>90820</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90820 1132 1240 +</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OrthoDB</td>\n",
       "      <td>366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.500000e-47</td>\n",
       "      <td>Transcript_100000</td>\n",
       "      <td>LAST</td>\n",
       "      <td>36</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>translated_nucleotide_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>\"Pfam:PF13465.2\"</td>\n",
       "      <td>homology:53130</td>\n",
       "      <td>zf-H2C2_2</td>\n",
       "      <td>Zinc-finger double domain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zf-H2C2_2 1 25 +</td>\n",
       "      <td>0.94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>198</td>\n",
       "      <td>127 201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.200000e-03</td>\n",
       "      <td>Transcript_100001</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>\"Pfam:PF13894.2\"</td>\n",
       "      <td>homology:53126</td>\n",
       "      <td>zf-C2H2_4</td>\n",
       "      <td>C2H2-type zinc finger</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zf-C2H2_4 2 23 +</td>\n",
       "      <td>0.94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>234</td>\n",
       "      <td>166 237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.000000e-03</td>\n",
       "      <td>Transcript_100001</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>\"Pfam:PF12171.4\"</td>\n",
       "      <td>homology:53128</td>\n",
       "      <td>zf-C2H2_jaz</td>\n",
       "      <td>Zinc-finger double-stranded RNA-binding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zf-C2H2_jaz 4 23 +</td>\n",
       "      <td>0.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>231</td>\n",
       "      <td>172 240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.600000e-03</td>\n",
       "      <td>Transcript_100001</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>\"Pfam:PF13912.2\"</td>\n",
       "      <td>homology:53133</td>\n",
       "      <td>zf-C2H2_6</td>\n",
       "      <td>C2H2-type zinc finger</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zf-C2H2_6 4 24 +</td>\n",
       "      <td>0.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>234</td>\n",
       "      <td>169 240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.800000e-03</td>\n",
       "      <td>Transcript_100001</td>\n",
       "      <td>HMMER</td>\n",
       "      <td>171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_hmm_match</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Dbxref               ID         Name  \\\n",
       "0               NaN  homology:425062        90820   \n",
       "1  \"Pfam:PF13465.2\"   homology:53130    zf-H2C2_2   \n",
       "2  \"Pfam:PF13894.2\"   homology:53126    zf-C2H2_4   \n",
       "3  \"Pfam:PF12171.4\"   homology:53128  zf-C2H2_jaz   \n",
       "4  \"Pfam:PF13912.2\"   homology:53133    zf-C2H2_6   \n",
       "\n",
       "                                      Note Parent              Target  \\\n",
       "0                                      NaN    NaN   90820 1132 1240 +   \n",
       "1                Zinc-finger double domain    NaN    zf-H2C2_2 1 25 +   \n",
       "2                    C2H2-type zinc finger    NaN    zf-C2H2_4 2 23 +   \n",
       "3  Zinc-finger double-stranded RNA-binding    NaN  zf-C2H2_jaz 4 23 +   \n",
       "4                    C2H2-type zinc finger    NaN    zf-C2H2_6 4 24 +   \n",
       "\n",
       "  accuracy bitscore database  end env_coords  phase         score  \\\n",
       "0      NaN      NaN  OrthoDB  366        NaN    NaN  5.500000e-47   \n",
       "1     0.94      NaN      NaN  198    127 201    NaN  1.200000e-03   \n",
       "2     0.94      NaN      NaN  234    166 237    NaN  4.000000e-03   \n",
       "3     0.93      NaN      NaN  231    172 240    NaN  2.600000e-03   \n",
       "4     0.96      NaN      NaN  234    169 240    NaN  5.800000e-03   \n",
       "\n",
       "               seqid source  start strand trunc                         type  \n",
       "0  Transcript_100000   LAST     36      +   NaN  translated_nucleotide_match  \n",
       "1  Transcript_100001  HMMER    126    NaN   NaN            protein_hmm_match  \n",
       "2  Transcript_100001  HMMER    168    NaN   NaN            protein_hmm_match  \n",
       "3  Transcript_100001  HMMER    171    NaN   NaN            protein_hmm_match  \n",
       "4  Transcript_100001  HMMER    171    NaN   NaN            protein_hmm_match  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
