{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. After tximport in R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Merge annotations for each species, connecting Trinity contigs/genes to annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "# requires dammit env\n",
    "# source activate dammit\n",
    "from dammit.fileio.gff3 import GFF3Parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "counts_matrices = \"/Users/johnsolk/Documents/UCDavis/Whitehead/counts_matrices/\"\n",
    "counts_files = os.listdir(counts_matrices)\n",
    "gene_out_dir = \"/Users/johnsolk/Documents/UCDavis/Whitehead/contig_gene_name_23June2018/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========\n",
      "F_heteroclitusMDPP\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(496133, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(59646, 10)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/johnsolk/anaconda3/lib/python3.6/site-packages/dammit/fileio/gff3.py:73: ParserWarning: Both a converter and dtype were specified for column attributes - only the converter will be used\n",
      "  dtype=dict(self.columns)):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(186798, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "38094\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22994\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16513\n",
      "Unique NCBI protein ID\n",
      "16513\n",
      "Unique annotated gene names\n",
      "19095\n",
      "Unique Trinity \"genes\"\n",
      "19095\n",
      "========\n",
      "F_parvapinis\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(279009, 9)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(37591, 9)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(126200, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "27660\n",
      "Unique Fhet gene names (one name per contig):\n",
      "20647\n",
      "Unique Fhet annotations, contigs with expression\n",
      "14641\n",
      "Unique NCBI protein ID\n",
      "14641\n",
      "Unique annotated gene names\n",
      "15643\n",
      "Unique Trinity \"genes\"\n",
      "15643\n",
      "========\n",
      "L_parva\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(275950, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(50621, 10)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(145529, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "34438\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22601\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16698\n",
      "Unique NCBI protein ID\n",
      "16698\n",
      "Unique annotated gene names\n",
      "18602\n",
      "Unique Trinity \"genes\"\n",
      "18602\n",
      "========\n",
      "F_notatus\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(286353, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(56636, 10)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(167061, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "36016\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22530\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16371\n",
      "Unique NCBI protein ID\n",
      "16371\n",
      "Unique annotated gene names\n",
      "18725\n",
      "Unique Trinity \"genes\"\n",
      "18725\n",
      "========\n",
      "F_heteroclitusMDPL\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(392026, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(54986, 10)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(188268, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "37378\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22849\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16534\n",
      "Unique NCBI protein ID\n",
      "16534\n",
      "Unique annotated gene names\n",
      "18980\n",
      "Unique Trinity \"genes\"\n",
      "18980\n",
      "========\n",
      "F_grandis\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(608505, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(63173, 10)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(234074, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "41637\n",
      "Unique Fhet gene names (one name per contig):\n",
      "24275\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16763\n",
      "Unique NCBI protein ID\n",
      "16763\n",
      "Unique annotated gene names\n",
      "19537\n",
      "Unique Trinity \"genes\"\n",
      "19537\n",
      "========\n",
      "F_zebrinus\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(194115, 5)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(64738, 5)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(102046, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "30569\n",
      "Unique Fhet gene names (one name per contig):\n",
      "20439\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16161\n",
      "Unique NCBI protein ID\n",
      "16161\n",
      "Unique annotated gene names\n",
      "18103\n",
      "Unique Trinity \"genes\"\n",
      "18103\n",
      "========\n",
      "A_xenica\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(248903, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(48046, 10)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(148980, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "33053\n",
      "Unique Fhet gene names (one name per contig):\n",
      "21806\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16064\n",
      "Unique NCBI protein ID\n",
      "16064\n",
      "Unique annotated gene names\n",
      "18016\n",
      "Unique Trinity \"genes\"\n",
      "18016\n",
      "========\n",
      "F_olivaceus\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(248630, 9)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(50728, 9)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(134207, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "31924\n",
      "Unique Fhet gene names (one name per contig):\n",
      "21600\n",
      "Unique Fhet annotations, contigs with expression\n",
      "15902\n",
      "Unique NCBI protein ID\n",
      "15902\n",
      "Unique annotated gene names\n",
      "17527\n",
      "Unique Trinity \"genes\"\n",
      "17527\n",
      "========\n",
      "F_catanatus\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(283212, 8)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(64577, 8)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(164596, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "36728\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22435\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16585\n",
      "Unique NCBI protein ID\n",
      "16585\n",
      "Unique annotated gene names\n",
      "19035\n",
      "Unique Trinity \"genes\"\n",
      "19035\n",
      "========\n",
      "F_sciadicus\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(178995, 5)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(59707, 5)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(98332, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "29127\n",
      "Unique Fhet gene names (one name per contig):\n",
      "20382\n",
      "Unique Fhet annotations, contigs with expression\n",
      "15972\n",
      "Unique NCBI protein ID\n",
      "15972\n",
      "Unique annotated gene names\n",
      "17766\n",
      "Unique Trinity \"genes\"\n",
      "17766\n",
      "========\n",
      "F_rathbuni\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(364231, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(62200, 10)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(176367, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "36768\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22718\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16410\n",
      "Unique NCBI protein ID\n",
      "16410\n",
      "Unique annotated gene names\n",
      "18395\n",
      "Unique Trinity \"genes\"\n",
      "18395\n",
      "========\n",
      "F_chrysotus\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(269787, 9)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(56554, 9)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(155206, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "35793\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22455\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16362\n",
      "Unique NCBI protein ID\n",
      "16362\n",
      "Unique annotated gene names\n",
      "18705\n",
      "Unique Trinity \"genes\"\n",
      "18705\n",
      "========\n",
      "F_diaphanus\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(288318, 7)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(55777, 7)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(128323, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "32584\n",
      "Unique Fhet gene names (one name per contig):\n",
      "21624\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16174\n",
      "Unique NCBI protein ID\n",
      "16174\n",
      "Unique annotated gene names\n",
      "18159\n",
      "Unique Trinity \"genes\"\n",
      "18159\n",
      "========\n",
      "L_goodei\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(266317, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(47556, 10)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(138485, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "33198\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22631\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16404\n",
      "Unique NCBI protein ID\n",
      "16404\n",
      "Unique annotated gene names\n",
      "18664\n",
      "Unique Trinity \"genes\"\n",
      "18664\n",
      "========\n",
      "F_nottii\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(123565, 3)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(95321, 3)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(69247, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "28419\n",
      "Unique Fhet gene names (one name per contig):\n",
      "18280\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16157\n",
      "Unique NCBI protein ID\n",
      "16157\n",
      "Unique annotated gene names\n",
      "19080\n",
      "Unique Trinity \"genes\"\n",
      "19080\n",
      "========\n",
      "F_similis\n",
      "========\n",
      "Number of Trinity \"genes\" (this is how we summarized expression):\n",
      "(392860, 10)\n",
      "Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.\n",
      "(55989, 10)\n",
      "Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):\n",
      "(154675, 8)\n",
      "Unique gene names in contigs with expression:\n",
      "34068\n",
      "Unique Fhet gene names (one name per contig):\n",
      "22338\n",
      "Unique Fhet annotations, contigs with expression\n",
      "16020\n",
      "Unique NCBI protein ID\n",
      "16020\n",
      "Unique annotated gene names\n",
      "18089\n",
      "Unique Trinity \"genes\"\n",
      "18089\n"
     ]
    }
   ],
   "source": [
    "for counts_file in counts_files:\n",
    "    if counts_file != \".DS_Store\":\n",
    "        species = counts_file.split(\"_\")[0]+\"_\"+counts_file.split(\"_\")[1]\n",
    "        print(\"========\")\n",
    "        print(species)\n",
    "        print(\"========\")\n",
    "        gene_out = gene_out_dir + species + \"_gene_counts_annotations_filt.csv\"\n",
    "        table = pd.read_csv(\"/Users/johnsolk/Documents/UCDavis/Whitehead/counts_matrices/\"+counts_file)\n",
    "        print('Number of Trinity \"genes\" (this is how we summarized expression):')\n",
    "        print(table.shape)\n",
    "        table = table.rename(columns={'Unnamed: 0': 'Gene'})\n",
    "        # Filter out genes with summ expression > 1 across each rows\n",
    "        #table_filt = table[table.sum(axis=1) > 1] \n",
    "        #table_filt = table[(table > 0).sum(axis=1) >= 5]\n",
    "        #https://stackoverflow.com/questions/34712050/filter-by-row-sum-and-value\n",
    "        table_filt = table [table.mask(table!=0).count(axis=1).div(float(len(table.columns))) < 0.05]\n",
    "        #print('Contigs with rowSum expression > 1')\n",
    "        print('Contigs with rowSum expression > 0 and also all rows that have 5% or more of its values equal to 0.')\n",
    "        print(table_filt.shape)\n",
    "        name = \"/Users/johnsolk/Documents/UCDavis/Whitehead/gff_annotations/\"+species+\".trinity_out.Trinity.fasta.dammit.gff3\"\n",
    "        conversion_contig = \"/Users/johnsolk/Documents/UCDavis/Whitehead/contig_gene_23June2018/\"+species+\"_contig_gene.csv\"\n",
    "        conversion_dammit = \"/Users/johnsolk/Documents/UCDavis/Whitehead/dammit_conversions/\"+species+\".trinity_out.Trinity.fasta.dammit.namemap.csv\"\n",
    "        annotations = GFF3Parser(filename=name).read()\n",
    "        annotations = annotations.dropna(subset=['Name'])\n",
    "        annotations[\"length\"] = annotations[\"end\"].subtract(annotations[\"start\"], fill_value=0)\n",
    "        pickonename = annotations.sort_values(by=['seqid', 'score'], ascending=True).query('score < 1e-05').drop_duplicates(subset='seqid')[['seqid', 'Name','Note','database','Dbxref','start','end','length']]\n",
    "        pickonename = pickonename.dropna(axis=0,how=\"all\")\n",
    "        print('Number of contigs with annotations (one annotation/contig, sorted by E-value < 1e-05 and picked the lowest):')\n",
    "        print(pickonename.shape)\n",
    "        conversions_dammit = pd.read_csv(conversion_dammit)\n",
    "        conversions_contig = pd.read_csv(conversion_contig)\n",
    "        conversions_dammit['Name'], conversions_dammit['info'] = conversions_dammit['original'].str.split(' ', 1).str\n",
    "        conversions_dammit = conversions_dammit[['Name','renamed']]\n",
    "        conversions_dammit.columns = ['Name','seqid']\n",
    "        coversions_contig = conversions_contig[['Name','Gene']]\n",
    "        merged_table = pd.merge(table_filt,conversions_contig,on=\"Gene\",how='left')\n",
    "        merged_table = pd.merge(merged_table,conversions_dammit,on=\"Name\",how='left')\n",
    "        merged_table = pd.merge(merged_table,pickonename,on=\"seqid\",how='inner')\n",
    "        print('Unique gene names in contigs with expression:')\n",
    "        print(len(merged_table.Name_y.unique()))\n",
    "        fhet = annotations[annotations['Name'].str.startswith(\"gi\")]\n",
    "        fhet_filtered = fhet.query('score < 1e-05').drop_duplicates(subset='seqid')[['seqid', 'Name','start','end','length']]\n",
    "        print(\"Unique Fhet gene names (one name per contig):\")\n",
    "        print(len(fhet_filtered.Name.unique()))\n",
    "        fhet_merged_table = pd.merge(merged_table,fhet_filtered,on='seqid',how='inner')\n",
    "        fhet_merged_table = fhet_merged_table.rename(columns = {'Gene':'TrinityGene','Name_x':'TrinityContig','seqid':'dammitSeqID','Name_y':'GeneName','start_x':'annotationStart','end_x':'annotationEnd','length_x':'annotationLength','Name':'FhetNCBIName','start_y':'FhetNCBIStart','end_y':'FhetNCBIEnd','length_y':'FhetNCBILength'})                                           \n",
    "        fhet_merged_table = fhet_merged_table.drop('Unnamed: 0', 1)\n",
    "        fhet_merged_table['split1'], fhet_merged_table['split2'],fhet_merged_table['split3'],fhet_merged_table['NCBIproteinID'],fhet_merged_table['NCBIproteinName'] = fhet_merged_table['FhetNCBIName'].str.split('|', 5).str\n",
    "        fhet_merged_table = fhet_merged_table.drop('split1',1)\n",
    "        fhet_merged_table = fhet_merged_table.drop('split2',1)\n",
    "        fhet_merged_table = fhet_merged_table.drop('split3',1)\n",
    "        fhet_merged_table = fhet_merged_table.sort_values(by=['TrinityGene','annotationLength'], ascending=False).drop_duplicates(subset='TrinityGene')\n",
    "        fhet_merged_table = fhet_merged_table.sort_values(by=['GeneName','annotationLength'], ascending=False).drop_duplicates(subset='GeneName')\n",
    "        print(\"Unique Fhet annotations, contigs with expression\")\n",
    "        print(len(fhet_merged_table.FhetNCBIName.unique()))\n",
    "        print('Unique NCBI protein ID')\n",
    "        print(len(fhet_merged_table.NCBIproteinID.unique()))\n",
    "        print('Unique annotated gene names')\n",
    "        print(len(fhet_merged_table.GeneName.unique()))\n",
    "        print('Unique Trinity \"genes\"')\n",
    "        print(len(fhet_merged_table.TrinityGene.unique()))\n",
    "        fhet_merged_table.to_csv(gene_out)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
